tfsdk/schema.go

```diff
@@ -208,3 +210,189 @@ func (s Schema) validate(ctx context.Context, req ValidateSchemaRequest, resp *V
 		})
 	}
 }
+
+// modifyAttributePlans runs all AttributePlanModifiers in all schema attributes
+func (s Schema) modifyAttributePlans(ctx context.Context, req ModifySchemaPlanRequest, resp *ModifySchemaPlanResponse) {
+	for name, a := range s.Attributes {
+		attrPath := tftypes.NewAttributePath().WithAttributeName(name)
+
+		attrPlan, diags := req.Plan.GetAttribute(ctx, attrPath)
+		resp.Diagnostics = append(resp.Diagnostics, diags...)
+		if diagnostics.DiagsHasErrors(diags) {
+			return
+		}
+
+		attributeReq := ModifyAttributePlanRequest{
+			AttributePath: attrPath,
+			Config:        req.Config,
+			State:         req.State,
+			Plan:          req.Plan,
+			ProviderMeta:  req.ProviderMeta,
+		}
+
+		attributeResp := &ModifyAttributePlanResponse{
+			AttributePlan: attrPlan,
+		}
+
+		a.modifyPlan(ctx, attributeReq, attributeResp)
+
+		if attributeResp.RequiresReplace {
+			resp.RequiresReplace = append(resp.RequiresReplace, attrPath)
+		}
+
+		resp.Diagnostics = append(resp.Diagnostics, attributeResp.Diagnostics...)
+
+		setAttrDiags := resp.Plan.SetAttribute(ctx, attrPath, attributeResp.AttributePlan)
+		resp.Diagnostics = append(resp.Diagnostics, setAttrDiags...)
+		if diagnostics.DiagsHasErrors(setAttrDiags) {
+			return
+		}
+
+		if a.Attributes != nil {
+			nm := a.Attributes.GetNestingMode()
+			switch nm {
+			case NestingModeList:
+				l, ok := attrPlan.(types.List)
+
+				if !ok {
+					err := fmt.Errorf("unknown attribute value type (%T) for nesting mode (%T) at path: %s", attrPlan, nm, attrPath)
+					resp.Diagnostics = append(resp.Diagnostics, &tfprotov6.Diagnostic{
+						Severity:  tfprotov6.DiagnosticSeverityError,
+						Summary:   "Attribute Plan Modification Error",
+						Detail:    "Attribute plan modifier cannot walk schema. Report this to the provider developer:\n\n" + err.Error(),
+						Attribute: attrPath,
+					})
+
+					return
+				}
+
+				for idx := range l.Elems {
+					for nestedName, nestedAttr := range a.Attributes.GetAttributes() {
```

Comment:
I think we want this to be recursive; right now, it looks like you only have one level of nesting supported. So a nested attribute inside a nested attribute inside a schema wouldn't have its plan modifiers run, as best I can tell.

If we made this method into a function that accepted a `map[string]Attribute` and an attribute path, we could just recursively call it for nested attributes:

```go
func modifyAttributesPlans(ctx context.Context, attrs map[string]Attribute, path *tftypes.AttributePath, req ModifySchemaPlanRequest, resp *ModifySchemaPlanResponse) {
```

Then this section could be something like

```go
for idx := range l.Elems {
  modifyAttributesPlans(ctx, a.Attributes.GetAttributes(), path.WithElementKeyInt(idx), req, resp)
  if diagnostics.DiagsHasErrors(resp.Diagnostics) {
    return
  }
}
```

Or at least that's what my vague imagining is suggesting it could look like :D

Reply:
This could also (maybe?) be resolved by doing the recursion on the `Attribute.modifyPlan` method, above.